drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c

```diff
@@ -0,0 +1,426 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/* @file
+ * @brief NRF Wi-Fi radio FICR programming functions
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr.h>
+#include "rpu_hw_if.h"
+#include "ficr_prog.h"
+
+
+LOG_MODULE_DECLARE(otp_prog, CONFIG_WIFI_LOG_LEVEL);
+
+static void write_word(unsigned int addr, unsigned int data)
+{
+	rpu_write(addr, &data, 4);
+}
+
+static void read_word(unsigned int addr, unsigned int *data)
+{
+	rpu_read(addr, data, 4);
+}
+
+unsigned int check_protection(unsigned int *buff, unsigned int off1, unsigned int off2,
+		    unsigned int off3, unsigned int off4)
+{
+	if ((buff[off1] == 0x00000000) &&
+		(buff[off2] == 0x00000000)  &&
+		(buff[off3] == 0x00000000) &&
+		(buff[off4] == 0x00000000))
+		return 0x00000000;
+	else if ((buff[off1] == 0xFFFFFFFF) &&
+		(buff[off2] == 0xFFFFFFFF) &&
+		(buff[off3] == 0xFFFFFFFF) &&
+		(buff[off4] == 0xFFFFFFFF))
+		return 0xFFFFFFFF;
+	else if ((buff[off1] == 0x50FA50FA) &&
+		(buff[off2] == 0x50FA50FA) &&
+		(buff[off3] == 0x50FA50FA) &&
+		(buff[off4] == 0x50FA50FA))
+		return 0x50FA50FA;
+	else
+		return 0xdeadbeef;
+}
+
+
+static void set_otp_timing_reg_40mhz(void)
+{
+	write_word(OTP_TIMING_REG1_ADDR, OTP_TIMING_REG1_VAL);
+
+	write_word(OTP_TIMING_REG2_ADDR, OTP_TIMING_REG2_VAL);
+}
+
+static int poll_otp_ready(void)
+{
+
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll != 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x4) == OTP_READY) {
+			return 0;
+		}
+		poll++;
+	}
+
+	LOG_ERR("OTP is not ready\n");
+	return -ENOEXEC;
+}
+
+
+static int req_otp_standby_mode(void)
+{
+	write_word(OTP_RWSBMODE_ADDR, 0x0);
+
+	return poll_otp_ready();
+}
+
+
+static int otp_wr_voltage_2V5(void)
+{
+	int err;
+
+	err = req_otp_standby_mode();
+
+	if (err) {
+		LOG_ERR("Failed Setting OTP voltage IOVDD to 2.5V\n");
+		return -ENOEXEC;
+	}
+	write_word(OTP_VOLTCTRL_ADDR, OTP_VOLTCTRL_2V5);
+	return 0;
+}
+
+static int poll_otp_read_valid(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll < 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x2) == OTP_READ_VALID) {
+			return 0;
+		}
+		poll++;
+	}
+	LOG_ERR("failed poll_otp_read_valid()\n");
+	return -ENOEXEC;
+}
+
+static int poll_otp_wrdone(void)
```

Comment:
poll_otp_wrdone(), poll_otp_read_valid(), poll_otp_ready() are very similar. Their code can be commonized.

Reply:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision